### ISSUE TITLE
Resolver's data not always passed to ActivatedRoute

### ISSUE THREAD

**Issue by maffelbaffel** [angular/angular#13946]
**I'm submitting a ...**  (check one with "x")
```
[x] bug report => search github for a similar issue or PR before submitting
[ ] feature request
[ ] support request => Please do not submit support request here, instead see https://github.com/angular/angular/blob/master/CONTRIBUTING.md#question
```

**Current behavior**
I've got a route with 2 configured resolvers:
````
{
     path: ':id',
     component: MyComponent,
     resolve: {
          one: ResolverOne,
          two: ResolverTwo
     }
}
````
and a component access data of both resolvers: 
````
@Component({
   ....
})
export class MyComponent implements OnInit {

    constructor(private route: ActivatedRoute) {
    }

    ngOnInit(): void {
        this.route.data.subscribe((data: {one: any, two: any}) => {
             // sometimes argument 'two' is undefined
            // thought both resolvers are being called before this code is executed!
             ...
        });
    }
}

````
Sometimes the second argument `two` is undefined. This happens only in 1/5 cases. The other 4 times it works correctly. Both resolvers are getting called every time!
My resolvers both look like this:
````
@Injectable()
export class OneResolver implements Resolve<any> {

    constructor(private someService: SomeService) {
    }

    resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<any> {
        let id = route.params['id'];

        // someService.get is returning a Observable here
        return this.someService.get(id);
    }
}

@Injectable()
export class ResolverTwo implements Resolve<any> {

    constructor(private someOtherService: SomeOtherService) {
    }

    resolve(route: ActivatedRouteSnapshot, state: RouterStateSnapshot): Observable<any> {
        let id = route.params['id'];

        // someOtherService.get is returning a Observable here
        return this.someOtherService.get(id);
    }
}
````

**Expected behavior**
Data of both Resolvers should be accessible in components after the route has been resolved.

**Please tell us about your environment:**
Windows

* **Angular version:** 2.4.3
Router: 3.4.3

* **Browser:**  Chrome 55.0.2883.87 m (64-bit), could not reproduce it in FF 49.0.2
 
* **Language:** TS: 2.0.10

* **Node (for AoT issues):**  6.9.1

**Comment by DzmitryShylovich** [comment]
Pls provide a plunkr with a small demo

**Comment by dbettini** [comment]
Can confirm this. Spent a lot of time trying to find what I was doing wrong, without even looking there. Components are loading after one of the resolvers get resolved, instead of waiting for all of them.

**Comment by maffelbaffel** [comment]
For me both resolvers seem to load before the component's observable 
````
this.route.data.subscribe((data: {one: any, two: any}) => {...});
````
gets resolved. I somehow just don't get both values. Only the `one`, but not `two` is available. 
What makes hard to debug is that its working most of the times. 

I will try to provide a working plunker later or maybe tomorrow, because i currently can't find the time doing one.

**Comment by DzmitryShylovich** [comment]
Sorry, but without a repro it's not actionable.

**Comment by dbettini** [comment]
Yes, my bad (it was before vacation that I ran into that problem, so I forgot). Yes, resolvers get resolved, both responses from the server are there before creating the component, but sometimes you get only one value.

**Comment by maffelbaffel** [comment]
Managed to create a simplified working plunker:
https://plnkr.co/edit/vtvN2KtPMDp69Sjx10io?p=preview
Check console output for resolver results.

**Comment by DzmitryShylovich** [comment]
btw does it mentioned anywhere that router supports more than one resolver for a single route?

**Comment by maffelbaffel** [comment]
Plunker didnt actually work :) just give me some more time.
not sure if it actually should. But it actually works (well sometimes)

**Comment by DzmitryShylovich** [comment]
> But it actually works (well sometimes)

will be hard to reproduce :(

**Comment by dbettini** [comment]
Your plunker is showing the bug for me. That is, resolver two is undefined

**Comment by DzmitryShylovich** [comment]
It's undefined because it should be `data.second` instead of `data.two`

**Comment by maffelbaffel** [comment]
yea but i had 
````
resolve: {
      one: FirstResolver,
      second: SecondResolver
    }
````

instead of 
````
resolve: {
      one: FirstResolver,
      two: SecondResolver
    }
````
2nd works.
I cannot get a example with a observable working because im not really into that plunker stuff...

**Comment by maffelbaffel** [comment]
in my application example the resolvers return a Observable, but i cannot get that working with plunker and SystemJs somehwo.

**Comment by DzmitryShylovich** [comment]
@maffelbaffel also works with observables https://plnkr.co/edit/zqcehseXN2KW3ylHGRYZ?p=preview

**Comment by maffelbaffel** [comment]
Well, i dunno why its not working for me then. This is mostly a simplified version of my applications code. Seems to be some mistake on my side then. 
Thanks alot for taking the time thought!

**Comment by DzmitryShylovich** [comment]
@maffelbaffel feel free to reopen if you will be able to reproduce :)

**Comment by dbettini** [comment]
Hello, I still had this problem (but couldn't repro it on JSBin) so I further investigated it myself. Basically, I was only getting one out of two properties on resolve, while both requests were already finished and available. Even weirder is, if I reverse the order of those two properties, it would work! 
In other words, this worked: 
```
 resolve: {
            objectiveNode: ObjectiveNodeResolve,
            moduleNode: ModuleNodeResolve,            
        }
```
while this resulted only in moduleNode being available:
```
 resolve: {
            moduleNode: ModuleNodeResolve,
            objectiveNode: ObjectiveNodeResolve,
        }
```
The resolver does a backend service call through a nodeResource created with [ngx-resource](https://github.com/troyanskiy/ng2-resource-rest). I also created a separate nodeService, which uses http service directly. I've tried all four combinations (using nodeService in ModuleNodeResolve while using nodeResource in ObjectiveNodeResolve, etc.). The results are following: for the first case, every combination worked. For the second case, those combinations that used nodeResource in ObjectiveNodeResolve always failed to have it available in route data. It started smelling like ngx-resource is at fault.

Tried delaying one of them at a time for a few seconds, and the component was always waiting for that amount of time to load, but still only had the same data as without the delay.

So by getting into angular router code a bit, I changed [this](https://github.com/angular/angular/blob/d58a242fe7b1d0e664733df3adfcbc458ec9dc67/packages/router/src/utils/collection.ts#L92) concatAll to mergeAll, and that resulted in everything working well. So, I realized those observables maybe aren't getting completed - that explains concatAll having values from only one stream, since if the first one isn't finishing, the second one will never be a part of the resulting stream, unlike with mergeAll. I confirmed this by patching the subscribe function of these observables with a third parameter, which was a "completed" function that should be called when the stream is done - it was never called.

Calling toPromise on observables that are returned from resolve methods fixed everything, and calls the "completed" function.

Then I found #7865. People seem to have a problem with the "completed" not firing, but doesn't seem to be reproducable on JSBin. That's happening here as well, but doesn't really affect the problem at question, because it never gets fired without using promises, but it still works sometimes. I then tried just calling first() on nodeResource.method(), but it errors with "no elements in sequence". So I tried Observable.of(nodeResource.method()) instead of using nodeResource.method().$observable. And that always works. It's also possible to call first() on it. 

So, TL;DR In my case, It was probably something with ngx-resource way of creating the observable. Not sure if @maffelbaffel found his problem, but anyway, for someone who finds this issue, try Observable.of() on your returning observable, or calling toPromise(). 
Still can't wrap my head around why order was important...

**Comment by urmichauhan** [comment]
let myblogid = this._route.snapshot.paramMap.get('blogid');
I have an array of object, each object has blogid, but above line not working, please solve this
